File: hss_okta_flutter/ios/Classes/HssOktaFlutterPlugin.swift
import Flutter
import OktaDirectAuth
import OktaOAuth2
import AuthFoundation
import UIKit
import WebAuthenticationUI




public class HssOktaFlutterPlugin: NSObject, FlutterPlugin,HssOktaFlutterPluginApi {

    let browserAuth = WebAuthentication.shared
    var flow : (any AuthenticationFlow)?
    var status : DirectAuthenticationFlow.Status?
    var deviceAuthorizationFlowContext : DeviceAuthorizationFlow.Context?
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        
        let messenger : FlutterBinaryMessenger = registrar.messenger()
        let api : HssOktaFlutterPluginApi & NSObjectProtocol = HssOktaFlutterPlugin.init()
        
        
        let signInFactory = WebSignInNativeViewFactory(messenger: registrar.messenger())
        registrar.register(signInFactory, withId: WebSignInNativeViewFactory.platformViewName)
        
        let signOutFactory = WebSignOutNativeViewFactory(messenger: registrar.messenger())
        registrar.register(signOutFactory, withId: WebSignOutNativeViewFactory.platformViewName)
        
        HssOktaFlutterPluginApiSetup.setUp(binaryMessenger: messenger, api: api)
        
    }
    
    func startDirectAuthenticationFlow(request: DirectAuthRequest, completion: @escaping (Result<AuthenticationResult?, Error>) -> Void) {
        
        
        if let config = try? OAuth2Client.PropertyListConfiguration(){
            flow = DirectAuthenticationFlow(issuer: config.issuer, clientId: config.clientId, scopes: config.scopes,supportedGrants: [.password,.otpMFA])
            
            Task{
                do{
                  
                    if let authFlow = flow as? DirectAuthenticationFlow{
                        
                        let status = try await authFlow.start(request.username, with: .password(request.password))
                        self.status = status
                        
                        switch status{
                        case .success(let token):
                            Credential.default = try Credential.store(token)
                            let userInfo = try await Credential.default?.userInfo()
                            completion(.success(constructAuthenticationResult(resultEnum: DirectAuthenticationResult.success, token: token, userInfo: userInfo)))
                           
                            
                        case .mfaRequired(_):
                            completion(.success(AuthenticationResult(result: DirectAuthenticationResult.mfaRequired)))
                    
                            
                        }
                    }else{
                        completion(.failure(HssOktaFlutterException(
                            code: "Error", message: "Incorrect flow",details: ""
                        )))
                    }
                    
                }catch let error{
                    completion(.failure(HssOktaFlutterException(
                        code: "Error", message: error.localizedDescription,details: ""
                    )))
                }
            }
        }else{
            
            completion(.failure(HssOktaFlutterException(
                code: "No .plist found", message: "Create a .plist containing okta configs",details: ""
            )))
        }
    }
    
    func continueDirectAuthenticationMfaFlow(otp: String, completion: @escaping (Result<AuthenticationResult?, Error>) -> Void) {
        Task{
            do{
                if let authFlow = flow as? DirectAuthenticationFlow{
                    status = try await authFlow.resume(self.status!, with: .otp(code: otp))
                    if case let .success(token) = status{
                        Credential.default = try Credential.store(token)
                        let userInfo = try await Credential.default?.userInfo()
                        
                        completion(.success(constructAuthenticationResult(resultEnum: DirectAuthenticationResult.success, token: token, userInfo: userInfo)))
                    }else{
                        completion(.failure(HssOktaFlutterException(
                            code: "Error", message:"failed to resume flow MFA failed",details: ""
                        )))
                    }
                }else{
                    completion(.failure(HssOktaFlutterException(
                        code: "Incorrect flow", message: "You have the incorrect flow",details: ""
                    )))
                }
            }catch let error{
                completion(.failure(HssOktaFlutterException(
                    code: "Error", message: error.localizedDescription, details: ""
                )))
            }
        }
    }
    
    func refreshDefaultToken(completion: @escaping (Result<Bool?, Error>) -> Void){
        
        
        if let credential = Credential.default{
            Task{
                do{
                    try await credential.refresh()
                }catch let error{
                    completion(.failure(HssOktaFlutterException(
                        code: "Error", message: error.localizedDescription,details: ""
                    )))
                }
            }
            completion(.success(true))
        }else{
            completion(.success(false))
        }
        
        
    }
        func revokeDefaultToken(completion: @escaping (Result<Bool?, Error>) -> Void){
            if let credential = Credential.default{
                Task{
                    do{
                        try await credential.revoke()
                        
                    }catch let error{
                        debugPrint(error)
                        completion(.success(false))
                    }
                }
                completion(.success(true))
            }else{
                completion(.success(false))
            }
        }
        
        func getCredential(completion: @escaping (Result<AuthenticationResult?, Error>) -> Void) {
            Task{
               
                do{
                    if let result = Credential.default{
                        let userInfo = try await result.userInfo()
                        
                        completion(.success(
                            constructAuthenticationResult(
                                resultEnum: nil,
                                token: result.token, userInfo: userInfo)))
                        
                    }else{
                        
                        completion(.failure(HssOktaFlutterException(
                            code: "No Credential found", message: "There is no default credential stored",details: ""
                        )))
                    }
                }catch let error{
                    completion(.failure(HssOktaFlutterException(
                        code: "Error", message: error.localizedDescription,details: ""
                    )))

                }
            }
        }
    
    func startDeviceAuthorizationFlow(completion: @escaping (Result<DeviceAuthorizationSession?, Error>) -> Void) {
        do{
            if (try? OAuth2Client.PropertyListConfiguration()) != nil{
                
                 flow = try DeviceAuthorizationFlow()
                
                Task{
                    do{
                        
                        if let authFlow = flow as? DeviceAuthorizationFlow{
                            
                            deviceAuthorizationFlowContext = try await authFlow.start()

                            completion(.success(DeviceAuthorizationSession(
                                userCode: deviceAuthorizationFlowContext?.userCode,
                                verificationUri: deviceAuthorizationFlowContext?.verificationUri.absoluteString
                            )))
                        }
                        
                        
                    }catch let error{
                        completion(.failure(HssOktaFlutterException(
                            code: "Error", message: error.localizedDescription,details: ""
                        )))
                    }
                }
            }
        }catch let error{
            completion(.failure(HssOktaFlutterException(
                code: "Error", message: error.localizedDescription,details: ""
            )))
        }
    }
    
    func resumeDeviceAuthorizationFlow(completion: @escaping (Result<AuthenticationResult?, Error>) -> Void) {
        
        Task{
            do{
                if let authFlow = flow as? DeviceAuthorizationFlow{
                    if let context = deviceAuthorizationFlowContext{
                        
                        let token = try await authFlow.resume(with: context)
                        Credential.default = try Credential.store(token)
                        
                        if let userInfo = try await Credential.default?.userInfo(){
                            completion(.success(constructAuthenticationResult(resultEnum: nil, token: token, userInfo: userInfo)))
                        }
                    }
                }
                
            }catch let error{
                
                completion(.failure(HssOktaFlutterException(
                    code: "Error", message: error.localizedDescription,details: ""
                )))
            }
        }
    }
   
    func startTokenExchangeFlow(deviceSecret: String, idToken: String, completion: @escaping (Result<AuthenticationResult?, Error>) -> Void) {
        Task{
            do{
                flow = try TokenExchangeFlow()
                
                if let authflow = flow as? TokenExchangeFlow{
                    let token = try await authflow.start(with: [.actor(type: .deviceSecret, value: deviceSecret),.subject(type: .idToken, value: idToken)])
                    
                    Credential.default = try Credential.store(token)
                    
                    if let result = Credential.default{
                       let userInfo = try await result.userInfo()
                        
                        completion(.success(
                            constructAuthenticationResult(
                                resultEnum: nil,
                                token: result.token, userInfo: userInfo)))
                        
                        
                        
                    }
                    
                    completion(.failure(HssOktaFlutterException(
                        code: "Error", message: "Failed saving credentails",details: ""
                    )))
                }
            }catch let error{
                completion(.failure(HssOktaFlutterException(
                    code: "Error", message: error.localizedDescription, details: ""
                )))
            }
            
        }
    }
    
    func getAllUserIds(completion: @escaping (Result<[String], Error>) -> Void) {
        Task{
            let ids = Credential.allIDs
            completion(.success(ids))
        }
    }
    
    func getToken(tokenId: String, completion: @escaping (Result<AuthenticationResult?, Error>) -> Void) {
        Task{
            do{
                if let fetchedCredential = try Credential.with(id: tokenId){
                    completion(.success(constructAuthenticationResult(
                        resultEnum: nil, token: fetchedCredential.token, userInfo: fetchedCredential.userInfo)))
                }
               
            }catch let error{
                completion(.failure(HssOktaFlutterException(
                    code: "Error", message: error.localizedDescription,details: ""
                )))
            }
        }
    }
    
   
    
    func removeCredential(tokenId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        Task{
            Task{
                do{
                    
                    try Credential.with(id:tokenId)?.remove()
                    completion(.success(true))
                    
                }catch let error{
                    completion(.failure(HssOktaFlutterException(
                        code: "Error", message: error.localizedDescription,details: ""
                    )))
                }
            }
        }
    }
    
    func setDefaultToken(tokenId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        Task{
            do{
               
                try Credential.tokenStorage.setDefaultTokenID(tokenId)
                completion(.success(true))
                
            }catch let error{
                completion(.failure(HssOktaFlutterException(
                    code: "Error", message: error.localizedDescription,details: ""
                )))
            }
        }}

    
    //    Helper methods
        
        func constructAuthenticationResult(resultEnum : DirectAuthenticationResult?, token : Token,userInfo : AuthFoundation.UserInfo?) -> AuthenticationResult{
            return AuthenticationResult(
                result: resultEnum,
                token: OktaToken(
                    id: token.id,
                    token: token.idToken?.rawValue ?? "",
                    issuedAt: Int64(((token.issuedAt?.timeIntervalSince1970 ?? 0) * 1000.0).rounded()),
                    tokenType: token.tokenType,
                    accessToken: token.accessToken,
                    scope: token.scope ?? "",
                    refreshToken: token.refreshToken ?? ""
                ),
                userInfo: UserInfo(userId: "", givenName: userInfo?.givenName ?? "", middleName: userInfo?.middleName ?? "", familyName: userInfo?.familyName ?? "", gender: userInfo?.gender ?? "", email: userInfo?.email ?? "", phoneNumber: userInfo?.phoneNumber ?? "", username: userInfo?.preferredUsername  ?? "")
            )
        }       
}

File: hss_okta_flutter/android/src/main/kotlin/dev/hypersense/software/hss_okta_flutter/HssOktaFlutterPlugin.kt
package dev.hypersense.software.hss_okta_flutter


import android.content.Context
import android.media.metrics.Event
import android.os.Build
import com.okta.authfoundation.claims.*
import com.okta.authfoundation.client.OidcClient
import com.okta.authfoundation.client.OidcClientResult
import com.okta.authfoundation.client.OidcConfiguration
import com.okta.authfoundation.client.dto.OidcUserInfo
import com.okta.authfoundation.credential.Credential
import com.okta.authfoundation.credential.CredentialDataSource
import com.okta.authfoundation.credential.CredentialDataSource.Companion.createCredentialDataSource
import com.okta.authfoundation.credential.RevokeTokenType
import com.okta.authfoundation.credential.Token
import com.okta.authfoundationbootstrap.*
import com.okta.oauth2.DeviceAuthorizationFlow
import com.okta.oauth2.DeviceAuthorizationFlow.Companion.createDeviceAuthorizationFlow
import io.flutter.embedding.engine.plugins.FlutterPlugin
import com.okta.oauth2.ResourceOwnerFlow.Companion.createResourceOwnerFlow
import com.okta.oauth2.TokenExchangeFlow
import com.okta.oauth2.TokenExchangeFlow.Companion.createTokenExchangeFlow
import dev.hypersense.software.hss_okta.AuthenticationResult
import dev.hypersense.software.hss_okta.DeviceAuthorizationSession
import dev.hypersense.software.hss_okta.DirectAuthRequest
import dev.hypersense.software.hss_okta.HssOktaFlutterPluginApi
import dev.hypersense.software.hss_okta.DirectAuthenticationResult
import dev.hypersense.software.hss_okta.HssOktaFlutterException
import dev.hypersense.software.hss_okta.OktaToken
import dev.hypersense.software.hss_okta.UserInfo
import io.flutter.plugin.common.EventChannel

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import okhttp3.HttpUrl.Companion.toHttpUrl

class HssOktaFlutterPlugin : HssOktaFlutterPluginApi, FlutterPlugin{
    private var context : Context? = null
    private val browserSignInChannelName = "dev.hypersense.software.hss_okta.channels.browser_signin";
    private val browserSignOutChannelName = "dev.hypersense.software.hss_okta.channels.browser_signout"

    private lateinit var browserSignInEventChannel : EventChannel
    private lateinit var browserSignOutEventChannel : EventChannel

    private lateinit var deviceAuthorizationFlow : DeviceAuthorizationFlow
    private lateinit var deviceAuthorizationFlowContext : DeviceAuthorizationFlow.Context

   private fun initializeOIDC() {
       println("Initializing OIDC Configuration")
       var credential: Credential

       val oidcConfiguration = OidcConfiguration(
           clientId =  BuildConfig.CLIENT_ID,
           defaultScope = BuildConfig.SCOPES,
       )

       println("${BuildConfig.ISSUER.toHttpUrl()}")

       val oidcClient = OidcClient.createFromDiscoveryUrl(
           oidcConfiguration,
           "${BuildConfig.ISSUER}/.well-known/openid-configuration".toHttpUrl(),)


       if(context == null){
           throw  Exception("Context is null")
       }

       CredentialBootstrap.initialize(oidcClient.createCredentialDataSource(context!!))

       println("Done Initializing OIDC Configuration /()")

       CoroutineScope(Dispatchers.IO).launch{
           credential = CredentialBootstrap.defaultCredential()
       }
   }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {

        HssOktaFlutterPluginApi.setUp(binding.binaryMessenger, this)

        browserSignInEventChannel = EventChannel(binding.binaryMessenger,browserSignInChannelName)
        browserSignOutEventChannel = EventChannel(binding.binaryMessenger,browserSignOutChannelName)

        binding.platformViewRegistry.registerViewFactory(
            WebSignInNativeViewFactory.platformViewName,
            WebSignInNativeViewFactory(browserSignInEventChannel)
        )

        binding.platformViewRegistry.registerViewFactory(
            WebSignOutNativeViewFactory.platformViewName,
            WebSignOutNativeViewFactory(browserSignOutEventChannel)
        )

        context = binding.applicationContext

        initializeOIDC()
    }
    
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        HssOktaFlutterPluginApi.setUp(binding.binaryMessenger, null)

    }


    override fun startDirectAuthenticationFlow(
        request: DirectAuthRequest,
        callback: (Result<AuthenticationResult?>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            val flow = CredentialBootstrap.oidcClient.createResourceOwnerFlow()

            when(val res = flow.start(request.username,request.password)){
                is OidcClientResult.Error -> {

                    callback.invoke(Result.failure(HssOktaFlutterException(
                        code = "Start error",
                        message = res.exception.message,
                        details = res.exception.stackTraceToString()
                    )))
                }
                is OidcClientResult.Success -> {

                    CredentialBootstrap.defaultCredential().storeToken(token = res.result)
                    var userInfo = CredentialBootstrap.defaultCredential().getUserInfo()
                    var userInfoResult = userInfo.getOrThrow()


                    callback.invoke(Result.success(composeOktaResult(res.result,userInfoResult)))
                }
            }
        }
    }
    
    override fun continueDirectAuthenticationMfaFlow(
        otp: String,
        callback: (Result<AuthenticationResult?>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                callback.invoke(Result.failure(HssOktaFlutterException(
                    code= "Error",
                    message ="Okta Android doesn't provide MFA OTP")))
            }catch (e: java.lang.Exception){

                callback.invoke(Result.failure(HssOktaFlutterException(
                    code = "Error",
                    message = e.message,
                    details = e.stackTraceToString()
                )))
            }
        }
    }

    override fun refreshDefaultToken(callback: (Result<Boolean?>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            when(val result = CredentialBootstrap.defaultCredential().refreshToken()){
                is OidcClientResult.Error ->   callback.invoke(Result.failure(
                    HssOktaFlutterException(
                    code = "Error",
                    message = result.exception.message,
                    details = result.exception.stackTraceToString()
                )))
                is OidcClientResult.Success -> callback.invoke(Result.success(true))
            }
        }
    }

    override fun revokeDefaultToken(callback: (Result<Boolean?>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            when(val result = CredentialBootstrap.defaultCredential().revokeToken(RevokeTokenType.ACCESS_TOKEN)){
                is OidcClientResult.Error -> callback.invoke(Result.failure(
                    HssOktaFlutterException(
                    code = "Error",
                    message = result.exception.message,
                    details = result.exception.stackTraceToString()
                )))
                is OidcClientResult.Success -> callback.invoke(Result.success(true))
            }
        }
    }

    override fun getCredential(callback: (Result<AuthenticationResult?>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            val credential = try{
                CredentialBootstrap.defaultCredential();
            }catch(e: java.lang.Exception){
                callback.invoke(Result.failure(HssOktaFlutterException(
                    code = "Error",
                    message = e.message,
                    details = e.stackTraceToString()
                )))
                null
            } ?: return@launch

            val userInfo = try{
                credential.getUserInfo().getOrThrow()
            }catch(e: java.lang.Exception){
                callback.invoke(Result.failure(HssOktaFlutterException(
                    code = "Error",
                    message = e.message,
                    details = e.stackTraceToString()
                )))
                null
            } ?: return@launch

            if(credential.token?.idToken == null){
                credential.delete()
                callback.invoke(Result.failure(HssOktaFlutterException(
                    code ="Token error",
                    message = "Previous token is empty and will be deleted")))
            }


            callback.invoke(Result.success(composeOktaResult(credential.token!!,userInfo)))
        }
    }

    override fun startDeviceAuthorizationFlow(callback: (Result<DeviceAuthorizationSession?>) -> Unit) {
        try {
        CoroutineScope(Dispatchers.IO).launch {
            deviceAuthorizationFlow = CredentialBootstrap.oidcClient.createDeviceAuthorizationFlow()

            when(val session = deviceAuthorizationFlow.start()){

                is OidcClientResult.Error -> {
                    callback.invoke(Result.failure(HssOktaFlutterException(
                        code = "Error",
                        message = session.exception.message,
                        details = session.exception.stackTraceToString()
                    )))
                }
                is OidcClientResult.Success -> {
                    deviceAuthorizationFlowContext = session.result

                    callback.invoke(Result.success(
                        DeviceAuthorizationSession(
                            session.result.userCode,
                            session.result.verificationUri
                        )
                    ))
                }
            }
        }
        }catch (exception : Exception){
            callback.invoke(Result.failure(HssOktaFlutterException(
                code = "Error",
                message = exception.message,
                details = exception.stackTraceToString()
            )))
        }
    }

    override fun resumeDeviceAuthorizationFlow(callback: (Result<AuthenticationResult?>) -> Unit) {
        try {
            CoroutineScope(Dispatchers.IO).launch {

                when(val session = deviceAuthorizationFlow.resume(deviceAuthorizationFlowContext)){
                    is OidcClientResult.Error -> {
                        callback.invoke(Result.failure(HssOktaFlutterException(
                            code = "Error",
                            message = session.exception.message,
                            details = session.exception.stackTraceToString()
                        )))
                    }
                    is OidcClientResult.Success -> {
                        CredentialBootstrap.defaultCredential().storeToken(session.result)

                        when(val userInfo = CredentialBootstrap.defaultCredential().getUserInfo()){
                            is OidcClientResult.Error ->  callback.invoke(Result.failure(HssOktaFlutterException(
                                code = "Error",
                                message = userInfo.exception.message,
                                details = userInfo.exception.stackTraceToString()
                            )))
                            is OidcClientResult.Success ->{
                                callback.invoke(Result.success(
                                    composeOktaResult(session.result,userInfo.result)
                                ))
                            }
                        }


                    }
                }
            }
        }catch (exception : Exception){
            callback.invoke(Result.failure(HssOktaFlutterException(
                code = "Error",
                message = exception.message,
                details = exception.stackTraceToString()
            )))
        }}

    override fun startTokenExchangeFlow(
        deviceSecret: String,
        idToken: String,
        callback: (Result<AuthenticationResult?>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            val credentialDataSource: CredentialDataSource = CredentialBootstrap.credentialDataSource

            val tokenExchangeFlow: TokenExchangeFlow = CredentialBootstrap.oidcClient.createTokenExchangeFlow()
            when(val result =tokenExchangeFlow.start(idToken, deviceSecret)){
                is OidcClientResult.Error -> callback.invoke(Result.failure(HssOktaFlutterException(
                    code = "Error",
                    message = result.exception.message,
                    details = result.exception.stackTraceToString()
                )))
                is OidcClientResult.Success ->{
                    CredentialBootstrap.defaultCredential().storeToken(token = result.result)

                    when(val userInfo = CredentialBootstrap.defaultCredential().getUserInfo()){
                        is OidcClientResult.Error ->  callback.invoke(Result.failure(HssOktaFlutterException(
                            code = "Error",
                            message = userInfo.exception.message,
                            details = userInfo.exception.stackTraceToString()
                        )))
                        is OidcClientResult.Success ->{
                            callback.invoke(Result.success(
                                composeOktaResult(result.result,userInfo.result)
                            ))
                        }
                    }
                }
            }
        }
    }

    override fun getAllUserIds(callback: (Result<List<String>>) -> Unit) {
       CoroutineScope(Dispatchers.IO).launch {
           try{
               val credentialList = CredentialBootstrap.credentialDataSource.listCredentials()
               val ids = credentialList.mapNotNull { credential -> credential.token?.idToken }

               callback.invoke(Result.success(ids))
           }catch (exception : Exception){
               callback.invoke(Result.failure(HssOktaFlutterException(
                   code = "Error",
                   message = exception.message,
                   details = exception.stackTraceToString()
               )))
           }

       }
    }

    override fun getToken(tokenId: String, callback: (Result<AuthenticationResult?>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            try{
                val credentialList = CredentialBootstrap.credentialDataSource.listCredentials()
                val selectedCredential = credentialList.firstOrNull { credential ->  credential.token?.idToken == tokenId}

               if(selectedCredential != null){
                   callback.invoke(Result.success(selectedCredential.token?.let {
                       composeOktaResult(
                           res = it,
                           userInfoResult = selectedCredential.getUserInfo().getOrThrow(),
                       )
                   }))
               }

                callback.invoke(Result.failure(HssOktaFlutterException(code = "Error","No token found")))
            }catch (exception : Exception){
                callback.invoke(Result.failure(HssOktaFlutterException(
                    code = "Error",
                    message = exception.message,
                    details = exception.stackTraceToString()
                )))
            }

        }
    }

    override fun removeCredential(tokenId: String, callback: (Result<Boolean>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            try{
                val credentialList = CredentialBootstrap.credentialDataSource.listCredentials()
                val selectedCredential = credentialList.firstOrNull { credential ->  credential.token?.idToken == tokenId}

                if(selectedCredential != null){
                    selectedCredential.delete()
                    callback.invoke(Result.success(true))
                }

                callback.invoke(Result.failure(
                    HssOktaFlutterException(
                    code = "Error",
                    message ="Failed to delete credential, Token not found")))

            }catch (exception : Exception){
                callback.invoke(Result.failure(HssOktaFlutterException(
                    code = "Error",
                    message = exception.message,
                    details = exception.stackTraceToString()
                )))
            }

        }
    }

    /// NOT AVAILABLE FOR ANDROID
    override fun setDefaultToken(tokenId: String, callback: (Result<Boolean>) -> Unit) {
       throw HssOktaFlutterException(code = "Missing", message = "This is not available for Android")
    }


    private fun composeOktaResult(res : Token, userInfoResult : OidcUserInfo) : AuthenticationResult{
        return AuthenticationResult(
            result = DirectAuthenticationResult.SUCCESS,
            token = OktaToken(
                id = "",
                issuedAt = userInfoResult.issuedAt?.toLong(),
                tokenType = res.tokenType,
                scope = res.scope,
                refreshToken = res.refreshToken,
                accessToken = res.accessToken,
                token = res.idToken
            ),
            userInfo = UserInfo(
                userId = userInfoResult.userId?:"",
                givenName = userInfoResult.givenName?:"",
                middleName = userInfoResult.middleName?:"",
                familyName = userInfoResult.familyName?:"",
                gender = userInfoResult.gender?:"",
                email = userInfoResult.email?:"",
                phoneNumber = userInfoResult.phoneNumber?:"",
                username = userInfoResult.username?:""
            )
        )
    }



}